arch/RISCV/RISCVDisassembler.c

```diff
@@ -0,0 +1,434 @@
+					     unsigned N) 
+{
+  	//assert(isUInt<N>(Imm) && "Invalid immediate");
+  	// Sign-extend the number in the bottom N bits of Imm after accounting for
+  	// the fact that the N bit immediate is stored in N-1 bits (the LSB is
+  	// always zero)
+  	//Inst.addOperand(MCOperand::createImm(SignExtend64<N>(Imm << 1)));
+  	MCOperand_CreateImm0(Inst, SignExtend64(Imm << 1, N));
+  	return MCDisassembler_Success;
+}
+
+static DecodeStatus decodeCLUIImmOperand(MCInst *Inst, uint64_t Imm,
+                                         int64_t Address,
+                                         const void *Decoder) 
+{
+  	//assert(isUInt<6>(Imm) && "Invalid immediate");
+  	if (Imm > 31) {
+    		Imm = (SignExtend64(Imm, 6) & 0xfffff);
+  	}
+  	//Inst.addOperand(MCOperand::createImm(Imm));
+  	MCOperand_CreateImm0(Inst, Imm);
+  	return MCDisassembler_Success;
+}
+
+static DecodeStatus decodeFRMArg(MCInst *Inst, uint64_t Imm,
+                                 int64_t Address,
+                                 const void *Decoder) 
+{
+  	//assert(isUInt<3>(Imm) && "Invalid immediate");
+  	if (!RISCVFPRndMode_isValidRoundingMode(Imm))
+    		return MCDisassembler_Fail;
+
+  	//Inst.addOperand(MCOperand::createImm(Imm));
+  	MCOperand_CreateImm0(Inst, Imm);
+  	return MCDisassembler_Success;
+}
+
+
+#include "RISCVGenDisassemblerTables.inc"
+
+static void clear_MI_insn_detail(MCInst *MI) 
+{
+  	if (MI->flat_insn->detail) {
+    		memset(MI->flat_insn->detail, 0, sizeof(cs_detail));
+  	}
+
+  	return;
+}
+
+// mark the load/store instructions through the opcode.
+static void markLSInsn(MCInst *MI, uint32_t in)
+{
+	/* 
+	   I   ld 0000011 = 0x03
+	       st 0100011 = 0x23
+	   F/D ld 0000111 = 0x07
+	       st 0100111 = 0x27
+	*/
+#define MASK_LS_INSN 0x0000007f
+	uint32_t opcode = in & MASK_LS_INSN;
+	if (0 == (opcode ^ 0x03) || 0 == (opcode ^ 0x07) ||
+	    0 == (opcode ^ 0x23) || 0 == (opcode ^ 0x27))
+		MI->flat_insn->detail->riscv.need_effective_addr = true;
+#undef MASK_LS_INSN
+	return;
+}
+
+static DecodeStatus RISCVDisassembler_getInstruction(int mode, MCInst *MI,
+				 const uint8_t *code, size_t code_len,
+				 uint16_t *Size, uint64_t Address,
+				 MCRegisterInfo *MRI) 
+{
+  	// TODO: This will need modification when supporting instruction set
+  	// extensions with instructions > 32-bits (up to 176 bits wide).
+  	uint32_t Inst = 0;
+  	DecodeStatus Result;
+
+  	// It's a 32 bit instruction if bit 0 and 1 are 1.
+  	if ((code[0] & 0x3) == 0x3) {
+      		if (code_len < 4) {
+        		*Size = 0;
+        		return MCDisassembler_Fail;
+      		}
+
+      		*Size = 4;
+      		// Get the four bytes of the instruction.
+      		//Encoded as little endian 32 bits.
+      		Inst = code[0] | (code[1] << 8) | (code[2] << 16) | ((uint32_t)code[3] << 24);
+		clear_MI_insn_detail(MI);
+		// Now we need mark what instruction need fix effective address output.
+    		if (MI->csh->detail) 
```

Comment:
This looks weird to me that you access the `detail` field after `clear_MI_insn_detail`

Reply:
looks like `init_MI_insn_detail` is a good name.
but whatever, just a name.

Reply:
Maybe you misunderstand, condition test at LINE 359, not the memory been cleaned LINE 357.

Reply:
> looks like init_MI_insn_detail is a good name.
> but whatever, just a name.
My fault, I misunderstood that `clear` is a destructor-like function, `init` will be much better.

Reply:
You are right, updated.